compiler/resolution/fcf-support.cpp

```diff
@@ -996,18 +775,425 @@ insertSharedParentFactory(const SharedFcfSuperInfo info,
+    auto ret = p != owner_;
+    return ret;
+  }
+
+  bool isSymbolOfInterest(Symbol* node) const {
+    auto ret = isLcnSymbol(node) && !isGlobal(node);
+    return ret;
+  }
+
+  void visitSymExpr(SymExpr* node) override {
+    auto sym = node->symbol();
+
+    if (!isOuterSymbol(sym) || !isSymbolOfInterest(sym)) return;
+
+    auto it = outerVariableToMentions_.find(sym);
+    if (it != outerVariableToMentions_.end()) {
+      it->second.push_back(node);
+      return;
+    }
+
+    outerVariables_.push_back(sym);
+    auto& v = outerVariableToMentions_[sym];
+    v.push_back(node);
+  }
+
+  bool enterFnSym(FnSymbol* node) override {
+    childFunctions_.push_back(node);
+    return false;
+  }
+};
 
-  return true;
+ClosureEnv::ClosureEnv(FnSymbol* owner) : owner_(owner) {
+  INT_ASSERT(owner);
+  auto p = owner->defPoint->parentSymbol;
+
+  INT_ASSERT(p->getModule() == owner->getModule());
+
+  // Visitor takes our fields by reference and populates them.
+  if (isFnSymbol(p) && p != owner->getModule()->initFn) {
+    auto v = OuterVariableCollector(owner_, outerVariables_,
+                                    outerVariableToMentions_,
+                                    childFunctions_);
+    owner->body->accept(&v);
+  }
+}
+
+FnSymbol* ClosureEnv::owner() const {
+  return owner_;
+}
+
+bool ClosureEnv::isEmpty() const {
+  bool ret = numOuterVariables() == 0;
+  return ret;
+}
+
+int ClosureEnv::numOuterVariables() const {
+  auto ret = ((int) outerVariables_.size());
+  return ret;
+}
+
+Symbol* ClosureEnv::outerVariable(int idx) const {
+  INT_ASSERT(0 <= idx && idx < numOuterVariables());
+  auto ret = outerVariables_[idx];
+  return ret;
+}
+
+int ClosureEnv::numMentions(Symbol* sym) const {
+  auto it = outerVariableToMentions_.find(sym);
+  if (it == outerVariableToMentions_.end()) return 0;
+  auto ret = ((int) it->second.size());
+  return ret;
+}
+
+SymExpr* ClosureEnv::firstMention(Symbol* sym) const {
+  INT_ASSERT(1 <= numMentions(sym));
+  auto ret = mention(sym, 0);
+  return ret;
+}
+
+SymExpr* ClosureEnv::mention(Symbol* sym, int idx) const {
+  auto it = outerVariableToMentions_.find(sym);
+  if (it == outerVariableToMentions_.end()) return nullptr;
+  int hi = ((int) it->second.size());
+  INT_ASSERT(0 <= idx && idx < hi);
+  auto ret = it->second[idx];
+  return ret;
+}
+
+const std::vector<FnSymbol*>& ClosureEnv::childFunctions() const {
+  return childFunctions_;
+}
+
+/************************************* | **************************************
+*                                                                             *
+*                                                                             *
+*                                                                             *
+************************************** | *************************************/
+
+static bool
+readConfigParamBool(ModuleSymbol* modSym, const char* configParamName,
+                    VarSymbol*& cachedValue) {
```

Comment:
The key part is that you have to make the params persist rather than have them be pruned from the tree after they are folded away. I think the rest is probably about the same amount of work.

I didn't want to add these as compiler flags because I didn't want to do all that plumbing for something fairly hidden. Also, I did this partly as a challenge for myself because everyone was suggesting to use the flags 😄.

I like to imagine that when `dyno` is the resolver it will be trivial to read the value of a param from the standard/internal modules. So this is just a stopgap until then.